File: leetcode20_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dic = {
            ')': '(',
            '}': '{',
            ']': '['
        }
        #validate check with stack
        for c in s:
            #if closed parentheses check the stack
            if not c in dic:
                stack.append(c)
            elif not stack or dic[c] != stack.pop():
                return False
        #check stack whether it's empty or not
        return len(stack) == 0
        

File: leetcode20_valid_parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let parens = { "[": "]", "{": "}", "(": ")" };
  let stack = [];
  for (let c of s) {
    if (c in parens) {
      stack.push(c);
    } else {
      if (stack.length === 0) return false;
      if (parens[stack.pop()] !== c) return false;
    }
  }
  if (stack.length !== 0) return false;
  return true;
};
